# Algoritmos de optimización - Seminario<br>
Nombre y Apellidos: Alvaro Sanchez Rocamora  <br>
Url Github: https://github.com/asancroc/03MIAR-Algoritmos-de-Optimizacion/tree/master/SEMINARIO<br>
Url Collab: https://colab.research.google.com/drive/1C8yWUBUS0C3ET15TZGRoJmP2LEJdZVcj?authuser=1#scrollTo=AiskKkzZfy4e<br>

Problema:
> 1. Sesiones de doblaje <br>
>2. Organizar los horarios de partidos de La Liga<br>
>3. Combinar cifras y operaciones


## Problema 3. Combinar cifras y operaciones

- El problema consiste en analizar el siguiente problema y diseñar un algoritmo que lo resuelva.
- Disponemos de las 9 cifras del 1 al 9 (excluimos el cero) y de los 4 signos básicos de las
operaciones fundamentales: suma(+), resta(-), multiplicación(*) y división(/)

- Debemos combinarlos alternativamente sin repetir ninguno de ellos para obtener una
cantidad dada. Un ejemplo sería para obtener el 4:
4+2-6/3*1 = 4

- Debe analizarse el problema para encontrar todos los valores enteros posibles planteando las
siguientes cuestiones:
    - ¿Qué valor máximo y mínimo se pueden obtener según las condiciones del problema?
    - ¿Es posible encontrar todos los valores enteros posibles entre dicho mínimo y máximo ?

(*) La respuesta es obligatoria





                                        

In [31]:
import itertools
from typing import List, Tuple, Set

In [32]:
def combinar_numeros_y_operadores(
    cifras: List[int], 
    operaciones: List[str],
    cant_objetivo: int
    ) -> Tuple[Set[str], float, float]:
    
    valor_min = 100000
    valor_max = -100000

    s_combinaciones = set()

    # Generamos todas las permutaciones de cifras y operaciones
    # teniendo en cuenta las restricciones del problema.
    # Cogemos todas las posiblemas combinaciones de 5 cifras
    for cifras_permutation in itertools.permutations(cifras, 5):
        # Cogemos todas las posiblemas combinaciones posibles de las 4 operaciones
        for operaciones_permutation in itertools.permutations(operaciones, 4):
            expression = ''
            
            # Mergeamos ambas listas para tener todas las combinaciones de num y cifras
            l_expresion = [val for pair in zip(cifras_permutation, operaciones_permutation) for val in pair]
            l_expresion += str(cifras_permutation[-1])

            expression = ''.join(str(val) for val in l_expresion)

            result = eval(expression)

            if result == cant_objetivo:
                s_combinaciones.add(expression)

            valor_min = min(valor_min, result)
            valor_max = max(valor_max, result)

    return s_combinaciones, valor_min, valor_max

In [68]:
cifras = [1, 2, 3, 4, 5, 6, 7, 8, 9]
operaciones = ['+', '-', '*', '/']
cant_objetivo = 4

combinaciones, v_minimo, v_maximo = combinar_numeros_y_operadores(
    cifras,
    operaciones,
    cant_objetivo
)

print(f"Numero de expresiones encontradas cuyo resultado es {cant_objetivo}: {len(combinaciones)}")
print(f"El valor más pequeño obtenido: {v_minimo}")
print(f"El valor más grande obtenido: {v_maximo}")
print("Las combinaciones resultantes:")
for expresion in combinaciones:
    print(expresion)

Numero de expresiones encontradas cuyo resultado es 4: 2112
El valor más pequeño obtenido: -70.71428571428571
El valor más grande obtenido: 78.83333333333333
Las combinaciones resultantes:
6-8+9*2/3
2-8+5/3*6
4/6*3-7+9
3/1+2*4-7
9+4/2-1*7
7+9-6*4/2
9+1-4/2*3
1*6/2-4+5
2*3/1-6+4
1-7+6/3*5
5-4+1*9/3
3+4*6/8-2
3-2/1*4+9
1*4/2+8-6
3+8*5/4-9
5+1-3/9*6
4+2*9/6-3
9-7+8*1/4
8-2*5+6/1
4*2-7+9/3
2+4-6/9*3
8/4*5+1-7
8+6-5/1*2
9/3-7+8*1
9-8+6/2*1
9-8+2*6/4
2-6/3+4*1
1+6/3*5-7
3-7+2*4/1
2-6/3+1*4
1*4-2+6/3
3-9+2/1*5
2+5-9*1/3
5+8-3/2*6
7-3/2*8+9
4*1+3-6/2
6/2+1*8-7
2-6/3*1+4
2-6/8*4+5
3*8/4-7+5
6-8+3/1*2
9+7-6*2/1
4-2+6/9*3
9+4/2-7*1
6+9*1/3-5
7-2*6/1+9
5+8/4-1*3
5*8/4-9+3
1+5-3/6*4
2*9/3+4-6
4-3*2/6+1
6+3*4/2-8
4/3*6+5-9
7+2/1*3-9
7/1+5-2*4
2*5+3-9/1
5-7+9*4/6
7+3*6/9-5
8/4-5+1*7
4*6/2+1-9
5*1+2-9/3
7/1-9+3*2
6*4/3-9+5
8*5/4-7+1
8*1/4-5+7
6+4/8*2-3
3-8*1/2+5
7-8/2*3+9
7+3/6*2-4
3+1*4-6/2
2-1+4*6/8
4/2-5+1*7
7-6*2/3+1
6/3*5+1-7
5-4+9*2/6
6*5/3+2-8
7-5+1/4*8
4+2-6*3/9
2/6*9+4-3
6+8-4*5/2
4/2-6+1*8
5

### **¿Qué valor máximo y mínimo se pueden obtener según las condiciones del problema?**
- El valor más pequeño obtenido: -70.71428571428571
- El valor más grande obtenido: 78.83333333333333

### **Es posible encontrar todos los valores enteros posibles entre dicho mínimo y máximo**
Se podría decir que es posible encontrar algunos valores enteros entre ellos. Sin embargo, es posible que no se puedan encontrar todos los valores enteros posibles en ese rango.

Dado que estamos trabajando con operaciones aritméticas y las cifras disponibles son números enteros, los resultados obtenidos podrían ser desde números enteros hasta fracciones con las cifras utilizadas.

No se garantiza que todos los valores enteros entre el mínimo y el máximo sean alcanzables.

Para confirmar si es posible encontrar todos los valores enteros posibles entre el mínimo y el máximo, sería necesario analizar exhaustivamente todas las combinaciones y evaluar los resultados obtenidos. Esto puede requerir una exploración exhaustiva del espacio de búsqueda, lo cual puede ser computacionalmente costoso y difícil de realizar en práctica.

In [69]:
# Ejemplo para ver si podemos encontrar solucion para todos los enteros entre el rango del minimo y maximo conseguido
cifras = [1, 2, 3, 4, 5, 6, 7, 8, 9]
operaciones = ['+', '-', '*', '/']
valor_min = -70
valor_max = 78

l_no_conseguibles = []

for cant_objetivo in list(range(valor_min, valor_max + 1)):
    combinaciones, _, _ = combinar_numeros_y_operadores(
        cifras,
        operaciones,
        cant_objetivo
    )
    
    if len(combinaciones) == 0:
        l_no_conseguibles.append(cant_objetivo)

print(f"Numero de enteros no conseguibles {len(l_no_conseguibles)}")
print("Estos ints no conseguibles son:")
for valor in l_no_conseguibles:
    print(valor)

Numero de enteros no conseguibles 2
Estos ints no conseguibles son:
-70
78


Con lo devuelto en la anterior llamado, podemos ver como con los valores -70 y 78 no somos campaces de lograr encomtrar combinación.

(*)¿Cuantas posibilidades hay sin tener en cuenta las restricciones?<br>



¿Cuantas posibilidades hay teniendo en cuenta todas las restricciones.




Respuesta

### **¿Cuantas posibilidades hay sin tener en cuenta las restricciones?**
El algoritmo implementado evalua exhaustivamente todas las combinaciones posibles de cifras y operadores respetando las restricciones de alternancia, utilizando al menos 5 números por expresión, junto a todos los operados, y utilizando los números del 1 al 9 sin repetirlos. Además, solo ha considerado las combinaciones que dan como resultado 4.

Por lo tanto, basándonos en los resultados del algoritmo, podemos concluir que existen 2112 posibilidades distintas para combinar cifras y operaciones con las restricciones mencionadas y obtener un resultado de 4.


### **¿Cuantas posibilidades hay teniendo en cuenta todas las restricciones?**

Si no tuviéramos las restricciones en el problema, la cantidad de posibilidades para combinar cifras y operaciones sería infinita. Sin las restricciones específicas, podríamos utilizar cualquier combinación de cifras y operadores de manera repetida o en cualquier orden para obtener el resultado deseado.

En este escenario sin restricciones, podríamos repetir una cifra o un operador tantas veces como quisiéramos, generar expresiones con longitudes variables y utilizar cualquier combinación de operadores y cifras disponibles. Esto daría lugar a un conjunto infinito de posibilidades para obtener un resultado determinado.

Modelo para el espacio de soluciones<br>
(*) ¿Cual es la estructura de datos que mejor se adapta al problema? Argumentalo.(Es posible que hayas elegido una al principio y veas la necesidad de cambiar, arguentalo)


Respuesta

### **¿Cual es la estructura de datos que mejor se adapta al problema? Argumentalo.(Es posible que hayas elegido una al principio y veas la necesidad de cambiar, arguentalo)**

Inicialmente, al abordar el problema de combinar cifras y operaciones, podría considerarse el uso de una estructura de datos como una lista para almacenar las cifras y otra lista para almacenar los operadores. Esto permite realizar la generación y combinación de las expresiones de manera más sencilla.

Sin embargo, al profundizar en el problema y considerando las restricciones adicionales, surge la necesidad de mantener un seguimiento de las combinaciones válidas que cumplen con el resultado objetivo. En este caso, una estructura de datos que se adapte bien al problema sería un conjunto (set) en lugar de una lista.

El uso de un set tiene varias ventajas:

1. Eliminación de duplicados: Al utilizar un conjunto, podemos eliminar automáticamente las expresiones duplicadas que podrían generarse durante el proceso de combinación. Esto es importante ya que no queremos contar múltiples veces la misma combinación válida.

2. Búsqueda eficiente: El conjunto tiene una estructura interna (hash table) que permite una búsqueda rápida y eficiente de elementos. Esto es útil para verificar si una combinación ya existe en el conjunto antes de agregarla, evitando así duplicados innecesarios.

3. Flexibilidad para la adición y eliminación de elementos: Un conjunto nos permite agregar nuevas combinaciones válidas de manera rápida y también eliminar elementos si se considera necesario.

Aunque inicialmente una lista podría ser considerada como una opción viable para almacenar las cifras y operadores, el uso de un conjunto se adapta mejor al problema debido a su capacidad para manejar de manera eficiente la eliminación de duplicados y la búsqueda rápida de elementos.

Según el modelo para el espacio de soluciones<br>
(*)¿Cual es la función objetivo?

(*)¿Es un problema de maximización o minimización?

Respuesta

### **¿Cual es la función objetivo?**
En el contexto del problema de combinar cifras y operaciones, el enunciado no especifica una función objetivo explícita. Sin embargo, se puede considerar que la función objetivo en este caso sería encontrar todas las expresiones válidas que cumplan con las restricciones y den como resultado el valor objetivo deseado (en este caso, el número 4). El objetivo es generar y almacenar todas las expresiones válidas que cumplan con estas condiciones.

Si tuvieramos como un orden fijo de colocación de los operados, podríamos obtener una función objetivo como la siguiente, por poner un ejemplo:

f(x, y, z, a, b) : ((x + y) - z) / a * b = 4

En esta función, x, y, z, a y b representan las cifras utilizadas en la combinación

### **¿Es un problema de maximización o minimización?**
El problema de combinar cifras y operaciones para obtener un resultado específico, como en este caso obtener 4, no es un problema de maximización ni minimización en el sentido tradicional.

No se trata de optimizar ninguna función objetivo en términos de maximizar o minimizar un valor específico. En cambio, se buscan todas las combinaciones posibles de cifras y operaciones que cumplan con las restricciones y den como resultado el número objetivo deseado.

El objetivo del problema es encontrar todas las combinaciones válidas, en lugar de buscar una solución óptima o encontrar la mejor solución en términos de maximización o minimización.

Diseña un algoritmo para resolver el problema por fuerza bruta

Respuesta

In [65]:
def combinar_numeros_y_operadores(
    cifras: List[int], 
    operaciones: List[str],
    cant_objetivo: int
    ) -> Tuple[Set[str], float, float]:
    
    valor_min = 100000
    valor_max = -100000

    s_combinaciones = set()

    # Generamos todas las permutaciones de cifras y operaciones
    # teniendo en cuenta las restricciones del problema.
    # Cogemos todas las posiblemas combinaciones de 5 cifras
    for cifras_permutation in itertools.permutations(cifras, 5):
        # Cogemos todas las posiblemas combinaciones posibles de las 4 operaciones
        for operaciones_permutation in itertools.permutations(operaciones, 4):
            expression = ''
            
            # Mergeamos ambas listas para tener todas las combinaciones de num y cifras
            l_expresion = [val for pair in zip(cifras_permutation, operaciones_permutation) for val in pair]
            l_expresion += str(cifras_permutation[-1])

            expression = ''.join(str(val) for val in l_expresion)

            result = eval(expression)

            if result == cant_objetivo:
                s_combinaciones.add(expression)

            valor_min = min(valor_min, result)
            valor_max = max(valor_max, result)

    return s_combinaciones, valor_min, valor_max

In [61]:
cifras = [1, 2, 3, 4, 5, 6, 7, 8, 9]
operaciones = ['+', '-', '*', '/']
cant_objetivo = 4

combinaciones, v_minimo, v_maximo = combinar_numeros_y_operadores(
    cifras,
    operaciones,
    cant_objetivo
)

print(f"Numero de expresiones encontradas cuyo resultado es {cant_objetivo}: {len(combinaciones)}")
print(f"El valor más pequeño obtenido: {v_minimo}")
print(f"El valor más pequeño obtenido: {v_maximo}")
print("Las combinaciones resultantes:")
for expresion in combinaciones:
    print(expresion)

Numero de expresiones encontradas cuyo resultado es 4: 2112
El valor más pequeño obtenido: -70.71428571428571
El valor más pequeño obtenido: 78.83333333333333
Las combinaciones resultantes:
6-8+9*2/3
2-8+5/3*6
4/6*3-7+9
3/1+2*4-7
9+4/2-1*7
7+9-6*4/2
9+1-4/2*3
1*6/2-4+5
2*3/1-6+4
1-7+6/3*5
5-4+1*9/3
3+4*6/8-2
3-2/1*4+9
1*4/2+8-6
3+8*5/4-9
5+1-3/9*6
4+2*9/6-3
9-7+8*1/4
8-2*5+6/1
4*2-7+9/3
2+4-6/9*3
8/4*5+1-7
8+6-5/1*2
9/3-7+8*1
9-8+6/2*1
9-8+2*6/4
2-6/3+4*1
1+6/3*5-7
3-7+2*4/1
2-6/3+1*4
1*4-2+6/3
3-9+2/1*5
2+5-9*1/3
5+8-3/2*6
7-3/2*8+9
4*1+3-6/2
6/2+1*8-7
2-6/3*1+4
2-6/8*4+5
3*8/4-7+5
6-8+3/1*2
9+7-6*2/1
4-2+6/9*3
9+4/2-7*1
6+9*1/3-5
7-2*6/1+9
5+8/4-1*3
5*8/4-9+3
1+5-3/6*4
2*9/3+4-6
4-3*2/6+1
6+3*4/2-8
4/3*6+5-9
7+2/1*3-9
7/1+5-2*4
2*5+3-9/1
5-7+9*4/6
7+3*6/9-5
8/4-5+1*7
4*6/2+1-9
5*1+2-9/3
7/1-9+3*2
6*4/3-9+5
8*5/4-7+1
8*1/4-5+7
6+4/8*2-3
3-8*1/2+5
7-8/2*3+9
7+3/6*2-4
3+1*4-6/2
2-1+4*6/8
4/2-5+1*7
7-6*2/3+1
6/3*5+1-7
5-4+9*2/6
6*5/3+2-8
7-5+1/4*8
4+2-6*3/9
2/6*9+4-3
6+8-4*5/2
4/2-6+1*8


Calcula la complejidad del algoritmo por fuerza bruta

Respuesta

### **Calcula la complejidad del algoritmo por fuerza bruta**
El tamaño del espacio de búsqueda está determinado por la cantidad de permutaciones posibles de las cifras y las operaciones. Dado que se consideran todas las permutaciones posibles de las 9 cifras y las 4 operaciones, el tamaño del espacio de búsqueda es 9! * 4! (factorial de 9 multiplicado por el factorial de 4).

El algoritmo utiliza dos bucles anidados para recorrer todas las permutaciones de cifras y operaciones. Dentro de estos bucles, se genera la expresión y se evalúa utilizando la función eval.

Dado que el algoritmo recorre todas las permutaciones posibles, la complejidad del algoritmo por fuerza bruta se puede estimar como O(9! * 4!) = O(n!).

Como el tamaño del espacio de búsqueda es 9! * 4! es constante, podemos simplificar la complejidad a O(1) ya que el tamaño del espacio de búsqueda no depende del tamaño de entrada.

Sin embargo, es importante tener en cuenta que la evaluación utilizando eval también tiene su propia complejidad, pero en este caso, la principal contribución a la complejidad total es el tamaño del espacio de búsqueda.

(*)Diseña un algoritmo que mejore la complejidad del algortimo por fuerza bruta. Argumenta porque crees que mejora el algoritmo por fuerza bruta

Respuesta

### **Diseña un algoritmo que mejore la complejidad del algortimo por fuerza bruta. Argumenta porque crees que mejora el algoritmo por fuerza bruta**
Es posible diseñar un algoritmo más eficiente que supere la complejidad temporal del enfoque de fuerza bruta en algunos casos.

El enfoque de fuerza bruta evalúa todas las combinaciones posibles, lo que implica una complejidad factorial. Sin embargo, en ciertos problemas, es posible aplicar técnicas de optimización o algoritmos más avanzados para reducir la complejidad temporal.

Por ejemplo, en el problema de combinar cifras y operaciones para obtener un resultado específico, se puede utilizar la programación dinámica o técnicas de búsqueda en espacio de estados para reducir la complejidad temporal.

La programación dinámica puede aprovechar subproblemas superpuestos y memorización para evitar recalcular soluciones ya conocidas, reduciendo así la cantidad total de cálculos requeridos. Esto puede conducir a una complejidad temporal más eficiente en comparación con el enfoque de fuerza bruta.

In [62]:
def combinar_numeros_y_operadores_dinam(cifras, operaciones, cant_objetivo):
    resultados = []

    # Creamos una matriz para almacenar los resultados
    dp = [[[] for _ in range(cant_objetivo + 1)] for _ in range(len(cifras) + 1)]

    # Recorremos la matriz
    for i in range(1, len(cifras) + 1):
        for j in range(1, cant_objetivo + 1):
            # Aplicamos las operaciones a las subexpresiones
            for operacion in operaciones:
                if operacion == '+' and j >= cifras[i - 1]:
                    for exp in dp[i][j - cifras[i - 1]]:
                        dp[i][j].append(exp + '+' + str(cifras[i - 1]))
                elif operacion == '-' and j + cifras[i - 1] <= cant_objetivo:
                    for exp in dp[i][j + cifras[i - 1]]:
                        dp[i][j].append(exp + '-' + str(cifras[i - 1]))
                elif operacion == '*' and j % cifras[i - 1] == 0 and j // cifras[i - 1] <= cant_objetivo:
                    for exp in dp[i][j // cifras[i - 1]]:
                        dp[i][j].append(exp + '*' + str(cifras[i - 1]))
                elif operacion == '/' and cifras[i - 1] != 0 and j * cifras[i - 1] <= cant_objetivo:
                    for exp in dp[i][j * cifras[i - 1]]:
                        dp[i][j].append(exp + '/' + str(cifras[i - 1]))

    # Obtenemos los resultados finales de la última celda de la matriz
    resultados = dp[len(cifras)][cant_objetivo]

    return resultados

In [63]:
cifras = [1, 2, 3, 4, 5, 6, 7, 8, 9]
operaciones = ['+', '-', '*', '/']
resultado_objetivo = 4

resultados = combinar_numeros_y_operadores_dinam(cifras, operaciones, resultado_objetivo)
print(resultados)

[]


(*)Calcula la complejidad del algoritmo

Respuesta

### **Calcula la complejidad del algoritmo**
Independientemente de que no haya conseguido hacer funcionar el problema, dejo por aquí la complejidad que tendría.

La complejidad temporal del algoritmo utilizando programación dinámica sería O(n * m * k) = O(n^3), donde n es la longitud de la lista de cifras, m es el resultado objetivo y k es el número de combinaciones válidas. 

Es importante destacar que aunque este algoritmo puede mejorar la eficiencia en comparación con el enfoque de fuerza bruta, aún tiene una complejidad alta debido a la naturaleza del problema y las restricciones establecidas.

Según el problema (y tenga sentido), diseña un juego de datos de entrada aleatorios

Respuesta

### **Según el problema (y tenga sentido), diseña un juego de datos de entrada aleatorios**
En este caso, no tendría mucho sentido crear ese conjunto de datos de entrada aleatorios, como mucho se podría ampliar o reducir el numero de valores y de operaciones.

Siguiendo esto, planteo usar de valores la siguiente lista [1, 4, 6, 7, 9] y de operaciones ['+', '-', '*'] para dar un valor final de 5.

Aplica el algoritmo al juego de datos generado



Respuesta

In [67]:
# Aplica el algoritmo al juego de datos generado

cifras = [1, 4, 6, 7, 9]
operaciones = ['+', '-', '*']
cant_objetivo = 5

combinaciones, v_minimo, v_maximo = combinar_numeros_y_operadores(
    cifras,
    operaciones,
    cant_objetivo
)

print(f"Numero de expresiones encontradas cuyo resultado es {cant_objetivo}: {len(combinaciones)}")
print(f"El valor más pequeño obtenido: {v_minimo}")
print(f"El valor más pequeño obtenido: {v_maximo}")
print("Las combinaciones resultantes:")
for expresion in combinaciones:
    print(expresion)

Numero de expresiones encontradas cuyo resultado es 5: 0
El valor más pequeño obtenido: 100000
El valor más pequeño obtenido: -100000
Las combinaciones resultantes:


Observando el conjunto de datos, restricciones para tener el valor 5 como resultado, vemos como el algoritmo de fuerza bruta no encuentra ninguna solución posible al problema.

Enumera las referencias que has utilizado(si ha sido necesario) para llevar a cabo el trabajo

Respuesta

1. https://stackoverflow.com/questions/65431247/how-to-get-all-mathematical-combinations-of-a-list-that-equal-a-result-number-in
2. https://docs.python.org/3/library/itertools.html

Describe brevemente las lineas de como crees que es posible avanzar en el estudio del problema. Ten en cuenta incluso posibles variaciones del problema y/o variaciones al alza del tamaño

Respuesta

### **Describe brevemente las lineas de como crees que es posible avanzar en el estudio del problema. Ten en cuenta incluso posibles variaciones del problema y/o variaciones al alza del tamaño**
1. Exploración de técnicas de optimización. Investigación en técnicas de optimización y algoritmos más eficientes que podrían aplicarse al problema. Algunas opciones pueden incluir técnicas de programación dinámica, memoización, búsqueda en espacio de estados, algoritmos de ramificación y poda

2. Pruebas y evaluación de los nuevos algoritmos. Realiza pruebas exhaustivas de los nuevos algoritmos utilizando diferentes conjuntos de datos y tamaños de entrada. Evaluando su rendimiento y verificando si cumple con las expectativas de mejora en la eficiencia

3. Escalabilidad y manejo de tamaños mayores. Si se desea abordar tamaños mayores del problema, es necesario evaluar la escalabilidad de los algoritmos propuestos. Considerando técnicas de optimización adicionales, como paralelización o distribución de tareas, para manejar tamaños de entrada más grandes